### PR TITLE
New Lint: avoid_using_api

### LINKED ISSUE #81: Avoid Using API Lint Rule
I've been working on a custom lint rule for banning external code by the identifier. Is this something you would want PR'd to this project?

I started working on it before custom_lint exposed the lint options, so I have some refactoring to do. I would rather add contribute it here since it is already setup for it.

### PR DESCRIPTION
**Pull request by getBoolean**
Closes #81

This lint has a lot of edge cases. Tests are included to verify basic functionality.

## Packages Added:

* [glob](https://pub.dev/packages/glob): Determines which files an entry should apply to (better than Regex for paths)
* [yaml](https://pub.dev/packages/yaml): To convert `YamlMap` to `Map` for the list of entries

## Usage

External code can be "deprecated" when there is a better option available:

```yaml
custom_lint:
  rules:
    - avoid_using_api:
      severity: info
      entries:
        - class_name: Future
          identifier: wait
          source: dart:async
          reason: "Future.wait should be avoided because it loses type safety for the results. Use a Record's `wait` method instead."
          severity: warning
```

Result:

```dart
void main() async {
  await Future.wait([...]); // LINT
  await (...,).wait; // OK
}
```

### Advanced Usage

Each entry also has `includes` and `excludes` parameters. These paths utilize [Glob](https://pub.dev/packages/glob) patterns to determine if a lint entry should be applied to a file.

For example, a lint to prevent usage of a domain-only package outside of the domain folder:

```yaml
custom_lint:
  rules:
    - avoid_using_api:
      severity: info
      entries:
        - source: package:modddels
          excludes:
            - "**/domain/**.dart"
          reason: "MODDDELS is only intended to be used in the domain layer."
```

### CODE REVIEW

File: lib/lints/banned_external_code/banned_external_code_rule.dart
Comment (yurii-prykhodko-solid) [inline lib/lints/banned_external_code/banned_external_code_rule.dart:130]: Use Platform.pathSeparator for Windows compat?

File: lib/utils/parameter_utils.dart
Comment (solid-vovabeloded) [inline lib/utils/parameter_utils.dart:130]: Please, use `Uri.normalizePath` instead.

File: lint_test/banned_external_code/id_ban/lib/id_ban.dart
Comment (yurii-prykhodko-solid) [inline lint_test/banned_external_code/id_ban/lib/id_ban.dart:26]:
```suggestion
  void test() {
    print('Hello World');
  }
  void useTest() {
    test(); // implicit `this`, no lint expected
  }
```

File: lib/lints/banned_external_code/banned_external_code_rule.dart
Comment (yurii-prykhodko-solid) [inline lib/lints/banned_external_code/banned_external_code_rule.dart:161]: Not sure these trampoline methods are useful. Maybe we should call the underlying function directly?

File: example/analysis_options.yaml
Comment (solid-vovabeloded) [inline example/analysis_options.yaml:22]: What do you think about renaming this lint rule to `avoid_using_api`?

File: lib/utils/parameter_utils.dart
Comment (yurii-prykhodko-solid) [inline lib/utils/parameter_utils.dart:121]: Could we dedupe a bit?
```suggestion
bool _matchesAnyGlob(List<String> globsList, String path) {
  final hasMatch =
      globsList.map(Glob.new).toList().any((glob) => glob.matches(path));
  return hasMatch;
}

bool _doesNotMatchGlobs(List<String> globList, String path) {
  return !_matchesAnyGlob(globList, path);
}
```

File: lib/lints/banned_external_code/models/banned_external_code_entry_parameters.dart
Comment (yurii-prykhodko-solid) [inline lib/lints/banned_external_code/models/banned_external_code_entry_parameters.dart:36]: IMO `source` should be required, and limited to external packages.

Whatever code is in the current package -- we can safely assume we have control of, and there already is a `@Deprecated` mechanism in place for cases where we want to phase out the usage of some internal API.

I'd rather not introduce an alternative way to do the same thing.

File: lint_test/banned_external_code/id_ban/analysis_options.yaml
Comment (yurii-prykhodko-solid) [inline lint_test/banned_external_code/id_ban/analysis_options.yaml:3]: Doesn't seem to work correctly otherwise.
```suggestion
    - custom_lint
```

File: example/analysis_options.yaml
Comment (yurii-prykhodko-solid) [reply example/analysis_options.yaml:22]: `deprecate_external_dependency`?

File: lint_test/banned_external_code/id_ban/analysis_options.yaml
Comment (getBoolean) [reply lint_test/banned_external_code/id_ban/analysis_options.yaml:3]: I'm pretty sure the issue is from [the analysis_options.yaml](https://github.com/invertase/dart_custom_lint/issues/148) file in the `lib/` folder. After deleting it the lints show up in the IDE

File: lib/lints/banned_external_code/banned_external_code_rule.dart
Comment (getBoolean) [reply lib/lints/banned_external_code/banned_external_code_rule.dart:161]: Yeah I agree is it a little weird, it's a holdover from before I refactored it to remove the duplicate code. I decided to keep it for the sake of consistency/readability.

If we make source required, these two can be removed entirely.

File: example/analysis_options.yaml
Comment (getBoolean) [reply example/analysis_options.yaml:26]: I think that would imply it has to be a member of a class, when it can also be the name of a method/variable in the global namespace of a source.

File: lib/lints/banned_external_code/banned_external_code_rule.dart
Comment (getBoolean) [reply lib/lints/banned_external_code/banned_external_code_rule.dart]: Is this what you had in mind?

```dart
    final Uri parentUri = p.toUri(parentSourcePath);
    final Uri sourceUri = p.toUri(sourceToMatch);

    final libraryName = parentUri.pathSegments.first;
    final sourceLibraryName = sourceUri.pathSegments.first;

    final matchesLibraryName = libraryName == sourceLibraryName;
    final shouldLintFileFromSource = sourceUri.pathSegments.length > 1;
    if (shouldLintFileFromSource) {
      final sourceFile = sourceUri.pathSegments.sublist(1).join('/');
      final libraryFile = parentUri.pathSegments.sublist(1).join('/');

      final matchesFile = p.equals(sourceFile, libraryFile);
      return matchesLibraryName && matchesFile;
    }
```

File: lib/lints/banned_external_code/banned_external_code_rule.dart
Comment (getBoolean) [reply lib/lints/banned_external_code/banned_external_code_rule.dart]: It was a bandaid fix, from before I intended to publish this. I've pushed a fix to the real issue, which was `ConstructorDeclaration` being linted (it should not lint declarations)

File: lib/utils/node_utils.dart
Comment (getBoolean) [reply lib/utils/node_utils.dart]: I've moved them to a new file in `banned_external_code` folder.

File: lib/utils/parameter_utils.dart
Comment (getBoolean) [reply lib/utils/parameter_utils.dart:49]: Sorry, this attribution is probably incorrect.

Upon further research, I think the correct source of this code is this [stackoverflow](https://stackoverflow.com/questions/23143653/how-to-load-yaml-file-to-the-map-in-dart) post, although it is not where I originally got it from. How should this be attributed?

A search for [YamlMapConverter on GitHub](https://github.com/search?q=YamlMapConverter&type=code&p=1) reveals many entries of this exact same code block with no attribution, so I was not sure what to put.

File: lib/utils/parameter_utils.dart
Comment (getBoolean) [reply lib/utils/parameter_utils.dart]: Removing `if (includeGlobs.isEmpty && excludeGlobs.isEmpty)` changes the behavior of the function. I am not sure about `isIncludeOnly` or `isExcludeOnly` yet, I need to setup a test for them.

This function is a little unintuitive, so I'll walk through it and convert the terminology to positives (this is mostly for my own understanding, it is a bit confusing)

* Line 95: If only include globs are given, only analyze the file if the path matches one of the include globs
* Line 99: If only exclude globs are given, only analyze the file if the path does not match any of the exclude globs
* Line 102-104: only analyze the file if the path matches one of the include globs and does not match any of the exclude globs

Maybe the function `shouldSkipFile` should be flipped and renamed to `shouldAnalyzeFile`?

File: lib/utils/parameter_utils.dart
Comment (getBoolean) [reply lib/utils/parameter_utils.dart]: Alright, I've also renamed it so it is more accurate to what it returns. For context, the globs fail when the drive letter is in the path so using relative is needed, and it doesn't make sense to match against folders above the project.
```dart
/// Converts path to relative using posix style and
/// replaces backslashes with forward slashes
String relativePath(String path, [String? root]) {
  final uriNormlizedPath = p.toUri(path).normalizePath().path;
  final uriNormlizedRoot =
      root != null ? p.toUri(root).normalizePath().path : null;

  final relative = p.posix.relative(uriNormlizedPath, from: uriNormlizedRoot);
  return relative;
}
```

File: example/analysis_options.yaml
Comment (getBoolean) [reply example/analysis_options.yaml:22]: I'll leave the final decision up to you and your team

File: lib/utils/parameter_utils.dart
Comment (getBoolean) [reply lib/utils/parameter_utils.dart]: Refactored version:

```dart
bool shouldSkipFile({
  required List<String> includeGlobs,
  required List<String> excludeGlobs,
  required String path,
  String? rootPath,
}) {
  final relative = relativePath(path, rootPath);
  final shouldAnalyzeFile =
      (includeGlobs.isEmpty || _matchesAnyGlob(includeGlobs, relative)) &&
          (excludeGlobs.isEmpty || _doesNotMatchGlobs(excludeGlobs, relative));
  return !shouldAnalyzeFile;
}

bool _matchesAnyGlob(List<String> globsList, String path) {
  final hasMatch =
      globsList.map(Glob.new).toList().any((glob) => glob.matches(path));
  return hasMatch;
}

bool _doesNotMatchGlobs(List<String> globList, String path) {
  return !_matchesAnyGlob(globList, path);
}
```

**Comment by yurii-prykhodko-solid** [comment]
@getBoolean Apologies for a tardy response -- I see you've made quite a bit of progress!

> 1. Rename parameter `id` to something that represents methods and variables, not sure what to name it.

Maybe `pattern`? You are using glob matching under the hood, seems consistent.

> 2. Make `source` required. What should it do if `source` is not provided?

I'd say just ignore it, and provide ample warning in rule documentation & usage examples.

> 3. Rename lint. I'll add `deprecate_external_api` as an additional choice

Oh this one sparked quite a discussion :D

The consensus was to use the `avoid_using_api` name.

File: lib/lints/banned_external_code/models/banned_external_code_entry_parameters.dart
Comment (yurii-prykhodko-solid) [reply lib/lints/banned_external_code/models/banned_external_code_entry_parameters.dart:36]: > I agree, though we can't stop users from specifying the source as package:my_app.

Well, if the users want to use a hacky loophole -- who are we to stop their creativity :)

**Comment by getBoolean** [comment]
@yurii-prykhodko-solid glob pattern matching is only used for includes/excludes filepaths, not the id/class name. Is that something we want to change?

**Comment by yurii-prykhodko-solid** [comment]
@getBoolean Ah, confused the two. Then `pattern` is no good.

`lexeme`? May be too broad in meaning.
`symbol`? Not really a common term in context of Dart lang.

And both of the above seem just as confusing as `id` to me. So I'd say we can leave it as is.

@solid-vovabeloded maybe you have some ideas?

**Comment by getBoolean** [comment]
Where should I put the lint documentation?

**Comment by yurii-prykhodko-solid** [comment]
@getBoolean 
> * Rename parameter `id` to something that represents methods and variables, not sure what to name it.

Discussed this internally, the consensus was to use `identifier`, with no abbreviation. "Identifier" is the term Dart team uses for everything named -- functions, variables, classnames, which seems consistent with what the lint is expecting.

> Where should I put the lint documentation?

That is an interesting question -- so far we've been maintaining this project as a drop-in solution, keeping configuration as minimal as it's possible, and therefore with little library user facing documentation.

I'd say we can leave this as an issue on this repo until we come up with a general approach of documenting our lints.

**Comment by getBoolean** [comment]
I think that's everything!

I do think documentation is needed, it will be very useful to know what all the lints are and what triggers them. Documentation for lints options can be included alongside that.

**Review by yurii-prykhodko-solid — COMMENTED** [review]
Looks good overall, great piece of work, @getBoolean!

Let's make the CI happy and we can merge.

**Comment by getBoolean** [comment]
I hope that is everything.

**Review by yurii-prykhodko-solid — APPROVED** [review]
Outstanding piece of work, @getBoolean!

**Comment by illia-romanenko** [comment]
Nice! Our CI still fails - it feels like we need to fix those issues before merging?

**Comment by getBoolean** [comment]
Not sure why its failing, I'll look into it when I get home from vacation in a week.

**Comment by getBoolean** [comment]
I've confirmed it working correctly on Windows and MacOS. It might be an issue somewhere in `AvoidUsingApiLinter._matchesSource` on Linux or CI causing the source check to fail.

**Comment by getBoolean** [comment]
Good thing I setup Windows/Linux dual boot already, made tracking down the issue a lot easier.

The issue was that `node.sourceUrl` would return the filepath on Linux for project files, while it was always the package path on Windows/MacOS. In the documentation, maybe we should specifically state that doing `source: package:myapp` will not work on Linux and should be avoided.

I also found two edge cases while debugging and fixed them

- lint for method usage on variable of a banned class type (source was previously ignored for it)
- lint for variable declaration of a banned class type if the type was inferred.

**Comment by yurii-prykhodko-solid** [comment]
> In the documentation, maybe we should specifically state that doing `source: package:myapp` will not work on Linux and should be avoided.

Which makes sense -- that's what 'Deprecated' is for.

Let's add this as a note in code documentation, I think the `AvoidUsingApiParameters` docs are the right place.

**Comment by yurii-prykhodko-solid** [comment]
> Let's add this as a note in code documentation, I think the AvoidUsingApiEntryParameters docs are the right place.

Done.

@getBoolean Let me know if you're happy with my changes, then we can finish this up.

**Comment by yurii-prykhodko-solid** [comment]
@getBoolean Looks like there's some more issues -- CI still in the red.

**Comment by getBoolean** [comment]
Before we merge, I want to investigate making `_matchesSource` take a non nullable parent source url, it was hiding a lint issue before.

@yurii-prykhodko-solid it looks like pub get needs to be run on those test packages in CI, is that possible?

**Comment by yurii-prykhodko-solid** [comment]
> @yurii-prykhodko-solid it looks like pub get needs to be run on those test packages in CI, is that possible?

Sure -- have a stab at our CI file. A `find -> exec` or a `find -> for` should do the trick.

**Comment by getBoolean** [comment]
Thanks, I think that is everything now

**Pull request merged**
